src/signals.c

```diff
@@ -401,7 +403,9 @@ static void sig_handle_timer(int signum)
 		event_ticksafety = event_pending = 1;
 	}
 
-#ifndef HAVE_MPI
+#ifdef HAVE_MPI
+	if (tty_has_keyboard())
+#else
```

Comment:
If all of this commit is about MPI stuff, then perhaps MPI should be mentioned in the commit message? Alternatively, if only some of this is about MPI, then maybe the commit should be split in two.

Reply:
The rest isn't. Actually this change "should" be the test for non-MPI runs as well, provided it doesn't give any regression on some systems and it might not be important enough to even profile.

Here's `tty_has_keyboard()`:
```c
int tty_has_keyboard(void)
{
#if !defined(__DJGPP__) && !defined(__MINGW32__) && !defined (_MSC_VER)
        return (tty_fd >= 0);
#else
        return !(in_stdin_mode && isatty(fileno(stdin)));
#endif
}
```
This could be very slightly worse than just `if (john_main_process)` (because of the function call, if nothing else) but in my tests it didn't matter enough to be measurable at all (on that platform).

Given that, I'm fine with just dropping this change. However, for some reason unclear to me today, MPI builds didn't even test for `john_main_process` so maybe we should instead drop the original `#ifndef` stuff and call it a day.

Reply:
We should also care about async signal safety here. Luckily, that `tty_has_keyboard()` function you quoted looks safe. We could want to add a comment near the function itself that we depend on its async signal safety by using it in `sig_handle_timer()` in MPI builds.